processing/src/main/java/org/apache/druid/segment/data/GenericIndexed.java

```diff
@@ -238,75 +232,243 @@ static int getNumberOfFilesRequired(int bagSize, long numWritten)
     return numberOfFilesRequired;
   }
 
+  protected final ObjectStrategy<T> strategy;
+  protected final boolean allowReverseLookup;
+  protected final int size;
 
-  private final boolean versionOne;
+  public GenericIndexed(
+      final ObjectStrategy<T> strategy,
+      final boolean allowReverseLookup,
+      final int size
+  )
+  {
+    this.strategy = strategy;
+    this.allowReverseLookup = allowReverseLookup;
+    this.size = size;
+  }
 
-  private final ObjectStrategy<T> strategy;
-  private final boolean allowReverseLookup;
-  private final int size;
-  private final ByteBuffer headerBuffer;
+  public abstract BufferIndexed singleThreaded();
 
-  private final ByteBuffer firstValueBuffer;
+  @Override
+  public abstract long getSerializedSize();
+
+  private static final class V1<T> extends GenericIndexed<T>
```

Comment:
super nit: this file seems like a bit of a mess (not really new, this change just exaggerates it a bit more), maybe these classes could live closer to `BufferIndexed` and after the shared instance methods of `GenericIndexed`?

Reply:
Yeah it has certainly grown complicated over time. Probably over-complicated. I am guessing you are still OK with merging this patch, given the ✅, but I would support efforts to make the class less complicated.